code/modules/power/overmap_solar.dm

```diff
@@ -0,0 +1,206 @@
+#define SOLAR_MAX_DIST 40
+
+// This is just to keep all the overmap solar stuff nice and tidy and not spread across three or four files.
+
+//
+// Overmap solars below. It'd be an absolutely pain to add a bunch of if statements and such to try and shoehorn normal function into them.
+//
+
+
+// SOLAR PANELS
+
+/obj/machinery/power/solar/overmap
+	var/obj/machinery/power/solar_control/overmap/overmap_control = null
+
+//calculates the fraction of the sunlight that the panel recieves
+/obj/machinery/power/solar/overmap/update_solar_exposure()
+	if(obscured)
+		sunfrac = 0
+		return
+	//find the smaller angle between the direction the panel is facing and the direction of the sun (the sign is not important here)
+	var/p_angle = min(abs(adir - overmap_control.overmap_tracker.get_sun_angle()), 360 - abs(adir - overmap_control.overmap_tracker.get_sun_angle()))
+
+	if(p_angle > 90)			// if facing more than 90deg from sun, zero output
+		sunfrac = 0
+
+		return
+	//Do distance checks from the star, keep sunfrac at a min of 0 and a max of 1.
+	var/obj/effect/overmap/star/star = locate(/obj/effect/overmap/star)
+	var/dist = get_dist(star, overmap_control.linked)
+	var/dist_penalty = dist * 0.1
+
+	sunfrac = cos(p_angle) ** 2
+	sunfrac -= (dist_penalty -= star.star_luminosity)
+	//Opacity check.
+	var/list/turflist = getline(overmap_control.linked, star)
+	for(var/turf/T in turflist)
+		var/obj/effect/overmap/overmap_effect = locate(/obj/effect/overmap) in T
+		if(overmap_effect && overmap_effect.opacity)
+			sunfrac = 0
+			break
+	if(!star.star_luminosity)
+		sunfrac = 0
+
+	sunfrac = Clamp(sunfrac, 0, 1)
+
+/obj/machinery/power/solar/overmap/unset_control()
+	if(overmap_control)
+		overmap_control.connected_panels.Remove(src)
+	overmap_control = null
+
+/obj/machinery/power/solar/overmap/Process()
+	if(stat & BROKEN)
+		return
+	if(!overmap_control) //if there's no sun or the panel is not linked to a solar control computer, no need to proceed
+		return
+
+	if(powernet)
```

Comment:
Should you also unset control if there's no powernet, or is the assumption that you don't need to unset if the powernet is already clear?